### PR TITLE
PlatformViewsController: clear composition_order_ in the beginning of each frame.

### PR DESCRIPTION
**Pull request by cyanglaz**
## Description

We used to clear composition_order_ in the end of the frame. This caused an issue with a frame drawn by [drawLastlayerTree](https://github.com/flutter/engine/blob/master/shell/common/rasterizer.cc#L147), because the endFrame is not called in such a frame. This causes a crash where a texture layer is constantly updating on top of a platform view, due to the same view_id is pushed to composition_order_ multiple times. (Again, because composition_order_ is not cleared)

This PR moves the composition_order_.clear() to the beginning of each frame.

This PR also removed FlutterViewController::EndFrame as it is not used anymore.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/70577

## Tests

I added the following tests:

testPlatformViewWithContinuousTexture

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Reviewer Checklist

- [x] I have submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

### CODE REVIEW

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm [hunk 1]
```diff
@@ -241,7 +241,8 @@
   gesture_recognizers_blocking_policies[idString] = gestureRecognizerBlockingPolicy;
 }
 
-void FlutterPlatformViewsController::SetFrameSize(SkISize frame_size) {
+void FlutterPlatformViewsController::BeginFrame(SkISize frame_size) {
+  composition_order_.clear();
```
Comment (blasten) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm:245]: fyi: Android does this: https://github.com/flutter/engine/blob/master/shell/platform/android/external_view_embedder/external_view_embedder.cc#L265.

File: testing/scenario_app/ios/Scenarios/Scenarios/ContinuousTexture.m [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+- (CVPixelBufferRef _Nullable)copyPixelBuffer {
+  return [self pixelBuffer];
+}
+
+- (CVPixelBufferRef)pixelBuffer
+{
+    NSDictionary *options = @{
+                              (NSString*)kCVPixelBufferCGImageCompatibilityKey : @YES,
```
Comment (blasten) [inline testing/scenario_app/ios/Scenarios/Scenarios/ContinuousTexture.m:48]: do you mind adding some comments explaining why this is all needed?

File: testing/scenario_app/lib/src/platform_view.dart [hunk 3]
```diff
@@ -613,6 +613,24 @@ class PlatformViewForTouchIOSScenario extends Scenario
...
 
+/// A simple platform view for testing platform view with a continuous texture layer.
+class PlatformViewWithContinuousTexture extends PlatformViewScenario {
+  /// Constructs a platform view with continuous texture layer
+  PlatformViewWithContinuousTexture(PlatformDispatcher dispatcher, String text, { int id = 0 })
+      : super(dispatcher, text, id: id);
+
+  @override
+  void onBeginFrame(Duration duration) {
+    print('begin frame');
```
Comment (blasten) [inline testing/scenario_app/lib/src/platform_view.dart:624]: is this intentional?

File: testing/scenario_app/lib/src/platform_view.dart [hunk 4]
```diff
@@ -613,6 +613,24 @@ class PlatformViewForTouchIOSScenario extends Scenario
...
+    final SceneBuilder builder = SceneBuilder();
+
+    builder.pushOffset(0, 0);
```
Comment (blasten) [inline testing/scenario_app/lib/src/platform_view.dart:627]: nit: this is a noop

File: testing/scenario_app/lib/src/platform_view.dart [hunk 5]
```diff
@@ -613,6 +613,24 @@ class PlatformViewForTouchIOSScenario extends Scenario
   }
 }
 
+/// A simple platform view for testing platform view with a continuous texture layer.
```
Comment (blasten) [inline testing/scenario_app/lib/src/platform_view.dart:616]: what does continuous texture means? Does it mean a texture that simulates a video being played?

Comment (blasten) [inline testing/scenario_app/lib/src/platform_view.dart:616]: what does continuous texture mean? Does it mean a texture that simulates a video being played?

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm (hunk 1 above)
Comment (blasten) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm:245]: could we also have a unit test for this?

**Comment by cyanglaz** [comment]
@blasten updated per comments. PTAL

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm (hunk 1 above)
Comment (blasten) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm:245]: could we make a`Reset()` method, similar to Android?  Why not also clearing `picture_recorders_`?

**Comment by mehmetf** [comment]
Folks, since this is a regression, let's try to get it in today to see whether we can land it internally by end of the week. It will be tight if at all possible.

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViewsTest.mm [hunk 6]
```diff
@@ -869,6 +869,61 @@ - (void)testFlutterPlatformViewControllerSubmitFrameWithoutFlutterViewNotCrashin
...
+
+  // First frame, |GetCurrentCanvases| is not empty after composite.
+  flutterPlatformViewsController->BeginFrame(SkISize::Make(300, 300));
+  flutter::MutatorsStack stack;
+  SkMatrix finalMatrix;
+  auto embeddedViewParams1 =
+      std::make_unique<flutter::EmbeddedViewParams>(finalMatrix, SkSize::Make(300, 300), stack);
+  flutterPlatformViewsController->PrerollCompositeEmbeddedView(0, std::move(embeddedViewParams1));
+  flutterPlatformViewsController->CompositeEmbeddedView(0);
+  XCTAssertEqual(flutterPlatformViewsController->GetCurrentCanvases().size(), (unsigned long)1);
```
Comment (blasten) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViewsTest.mm:912]: does `1UL` works here?

Comment (blasten) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViewsTest.mm:912]: does `1UL` work here?

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViewsTest.mm [hunk 7]
```diff
@@ -869,6 +869,61 @@ - (void)testFlutterPlatformViewControllerSubmitFrameWithoutFlutterViewNotCrashin
...
+  flutterPlatformViewsController->BeginFrame(SkISize::Make(300, 300));
+  XCTAssertEqual(flutterPlatformViewsController->GetCurrentCanvases().size(), (unsigned long)0);
+
+  auto embeddedViewParams2 =
+      std::make_unique<flutter::EmbeddedViewParams>(finalMatrix, SkSize::Make(300, 300), stack);
+  flutterPlatformViewsController->PrerollCompositeEmbeddedView(0, std::move(embeddedViewParams2));
+  flutterPlatformViewsController->CompositeEmbeddedView(0);
+  XCTAssertEqual(flutterPlatformViewsController->GetCurrentCanvases().size(), (unsigned long)1);
+
+  flutterPlatformViewsController->Reset();
```
Comment (blasten) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViewsTest.mm:924]: Is this needed?

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViewsTest.mm [hunk 8]
```diff
@@ -869,6 +869,61 @@ - (void)testFlutterPlatformViewControllerSubmitFrameWithoutFlutterViewNotCrashin
...
+  SkMatrix finalMatrix;
+  auto embeddedViewParams1 =
+      std::make_unique<flutter::EmbeddedViewParams>(finalMatrix, SkSize::Make(300, 300), stack);
+  flutterPlatformViewsController->PrerollCompositeEmbeddedView(0, std::move(embeddedViewParams1));
+  flutterPlatformViewsController->CompositeEmbeddedView(0);
+  XCTAssertEqual(flutterPlatformViewsController->GetCurrentCanvases().size(), (unsigned long)1);
+
+  // Second frame, |GetCurrentCanvases| should be empty at the start
+  flutterPlatformViewsController->BeginFrame(SkISize::Make(300, 300));
+  XCTAssertEqual(flutterPlatformViewsController->GetCurrentCanvases().size(), (unsigned long)0);
```
Comment (blasten) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViewsTest.mm:916]: nit: `XCTAssertTrue(flutterPlatformViewsController->GetCurrentCanvases().empty())`

File: testing/scenario_app/ios/Scenarios/Scenarios/ContinuousTexture.m [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+  void* pxdata = CVPixelBufferGetBaseAddress(pxbuffer);
+
+  CGColorSpaceRef rgbColorSpace = CGColorSpaceCreateDeviceRGB();
+  CGContextRef context = CGBitmapContextCreate(pxdata, width, height, 8, 4 * width, rgbColorSpace,
+                                               kCGImageAlphaNoneSkipFirst);
+  NSParameterAssert(context);
+
+  CGContextDrawImage(context, CGRectMake(0, 0, width, height), self.image.CGImage);
+  CGColorSpaceRelease(rgbColorSpace);
+  CGContextRelease(context);
```
Comment (blasten) [inline testing/scenario_app/ios/Scenarios/Scenarios/ContinuousTexture.m:74]: What parts of this method are essential to test and why?

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViewsTest.mm (hunk 6 above)
Comment (cyanglaz) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViewsTest.mm]: yes, updated

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViewsTest.mm (hunk 7 above)
Comment (cyanglaz) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViewsTest.mm]: no, removed

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViewsTest.mm (hunk 8 above)
Comment (cyanglaz) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViewsTest.mm]: done!

File: testing/scenario_app/ios/Scenarios/Scenarios/ContinuousTexture.m (hunk 9 above)
Comment (cyanglaz) [reply testing/scenario_app/ios/Scenarios/Scenarios/ContinuousTexture.m]: This is not required. removed.

**Review by blasten — APPROVED** [review]
LGTM

**Review by iskakaushik — APPROVED** [review]
LGTM

**Pull request merged**